### PR TITLE
Replace integers with explicit integers (1 -> 1L, etc.)

### PR DESCRIPTION
**Pull request by MichaelChirico**
Part of #2572, tried to make sure I respected "strong typing" (don't force `integer` on potential `numeric`s).

Command to find these:

    #numeric ones (1 -> 1L)
    grep '^[^#]*[^a-zA-Z_]1[^L]' R/* > ~/Desktop/all1s.out
    #for all integers
    grep '^[^#]*[^a-zA-Z_0-9][0-9][^0-9L]' R/* > ~/Desktop/numeric_integers.out

### CODE REVIEW

**Comment by mattdowle** [comment]
Very nice!
I'm not sure the status of global internal small integer constants in base R (i.e. whether altrep got to that or not and if that's been merged to R-devel) but I know it was planned.  Either way, this PR is needed to have the best chance to benefit from that.
Interested to see the time of `test.data.table()` (without Suggests packages, just plain DT tests) before and after.  But even if no speedup on that, it's still worth doing.

**Comment by MichaelChirico** [comment]
`head` and `tail` `GForce` optimization failures are legit:

```
dt = data.table(
  x  = sample(letters, 300, TRUE),
  d1 = as.numeric(sample(-10:10, 300, TRUE)),
  d2 = as.numeric(sample(c(NA, NaN, -10:10), 300, TRUE))
)
# not optimized now
dt[x %in% letters[15:20], head(.SD,1), by=x, verbose=TRUE]
# still optimized
dt[x %in% letters[15:20], head(.SD,1L), by=x, verbose=TRUE]
#idem
dt[x %in% letters[15:20], tail(.SD,1), by=x, verbose=TRUE]
dt[x %in% letters[15:20], tail(.SD,1L), by=x, verbose=TRUE]
```

Doesn't make sense to require users to run `head(.SD, 1L)` to benefit from optimization (even if it's best practice); `GForce` should detect and convert.

Investigating, though I'm in the dark here for a moment. Sometimes strong typing would be nice 😬

**Comment by MichaelChirico** [comment]
Also, should probably turn the lines like:

    if (verbose) {cat(round(proc.time()[3L]-last.started.at, 3L),"secs\n");flush.console}

into a function since they're so common -- @mattdowle IIUC this is what the `timetaken` function is for?

File: R/data.table.R [hunk 1]
```diff
@@ -1687,12 +1687,12 @@ chmatch2 <- function(x, table, nomatch=NA_integer_) {
...
           cond = is.call(q) && as.character(q[[1L]]) %chin% gfuns && !is.call(q[[2L]])
-          ans  = cond && (length(q)==2 || identical("na",substring(names(q)[3L],1,2)))
+          ans  = cond && (length(q)==2L || identical("na",substring(names(q)[3L], 1L, 2L)))
           if (identical(ans, TRUE)) return(ans)
-          ans = cond && length(q)==3 && ( as.character(q[[1]]) %chin% c("head", "tail") &&
-                             (identical(q[[3]], 1) || identical(q[[3]], 1L)) ||
-                          as.character(q[[1]]) %chin% "[" && is.numeric(q[[3]]) &&
-                            length(q[[3]])==1 && q[[3]]>0 )
+          ans = cond && length(q)==3L && ( as.character(q[[1L]]) %chin% c("head", "tail") &&
+                             (identical(q[[3L]], 1L) || identical(q[[3L]], 1L)) ||
```
Comment (MichaelChirico) [inline R/data.table.R]: Any reason this can't just be `q[[3L]] == 1L`?

File: R/data.table.R [hunk 2]
```diff
@@ -1687,12 +1687,12 @@ chmatch2 <- function(x, table, nomatch=NA_integer_) {
...
+          ans  = cond && (length(q)==2L || identical("na",substring(names(q)[3L], 1L, 2L)))
           if (identical(ans, TRUE)) return(ans)
-          ans = cond && length(q)==3 && ( as.character(q[[1]]) %chin% c("head", "tail") &&
-                             (identical(q[[3]], 1) || identical(q[[3]], 1L)) ||
-                          as.character(q[[1]]) %chin% "[" && is.numeric(q[[3]]) &&
-                            length(q[[3]])==1 && q[[3]]>0 )
+          ans = cond && length(q) == 3L &&
+            length(q[[3L]]) == 1L && is.numeric(q[[3L]]) && (
+              as.character(q[[1L]]) %chin% c("head", "tail") && q[[3L]] == 1L ||
+                as.character(q[[1L]]) %chin% "["  && q[[3L]] > 0 )
```
Comment (MichaelChirico) [inline R/data.table.R]: The logic now seems more natural to me (in particular, I see no reason not to check `is.numeric` and `length` conditions _only_ for the `[` case and not for the `head`/`tail` case -- perhaps a few tests are in order?), and the code reads better too.

Comment (mattdowle) [reply R/data.table.R]: The use of `identical` may have had `NA` in mind. For the `head / tail` case, `is.numeric` and `length` were being checked via `identical()`. Yes more tests would be great.  Seems like codecov agrees and is not-passing which is good (confirms it's a good exercise).
The line needs a comment saying what it's supposed to be doing, doesn't it!
Can you remove the added spaces around `==` please.  To my eyes there is now lots of separate things all with space between them and it's harder to see the tree of logic.  `=` is already a different character to letters (and is not valid in symbols) so `=` is already a visual separator.  Removing the spaces around `==` allows you to budge the pieces together in this case, so that the remaining spaces are more helpful.
Isn't head and tail with `n>1` now going to pass?  That doesn't seem to be the intent of the original line (only `n==1L` and `n==1` were allowed, iiuc).
An extra set of parens around the left and right of the last `||` might make it clearer?  There was a paren around the left of that `||` in the original but not on its right.

**Comment by MichaelChirico** [comment]
@mattdowle Ran this to test speed:

```
library(data.table)
library(magrittr)
replicate(10, {
  capture.output(suppressMessages(test.data.table())) %>% 
    capture.output %>% 
    grep('completed ok', ., value = TRUE) %>%
    gsub('.*completed ok in ([0-9:]+).*', '\\1', .) %>%
    strsplit(':') %>% extract2(1L) %>% extract(-1L) %>%
    as.numeric %>% (function(x) {x[1L] + x[2L]/60})
}) %>% mean
```

Results from testing on version currently at `master`: 

    # [1] 1.818333

Compared to this branch:

    # [1] 1.758333

About a 5% improvement... pretty damn good if you ask me! Certainly more than I expected.

(I do have many of the external packages; perhaps we should add a `testDeps` argument to `test.data.table`?)

```
Loading required package: bit64
Loading required package: bit
Loading required package: knitr
Loading required package: ggplot2
Loading required package: plyr
Loading required package: reshape2
Loading required package: nlme
Loading required package: curl
Loading required package: rmarkdown
Loading required package: parallel
```

**Comment by MichaelChirico** [comment]
Certainly easier to do, but I wasn't sure if there's any overhead in `test.data.table()` that's skipped by its internal timer

**Comment by mattdowle** [comment]
I'd be surprised if the overhead is more than negligible.  Easy to check.

**Comment by MichaelChirico** [comment]
Hmm, slightly different for me:

```
All 7661 tests in inst/tests/tests.Rraw completed ok in 00:01:48 on Thu Jan 18 16:46:12 2018
   user  system elapsed 
166.901   7.128 109.743 
```

anyway i'll run with `microbenchmark`

**Comment by MichaelChirico** [comment]
Not as impressive this time (<1%):

```
Unit: seconds
            expr      min       lq     mean   median       uq     max neval
 explicitInteger  96.4129  97.7939 101.9369 101.4272 105.4168 108.409    10
          master 100.3113 100.9985 102.4299 102.3071 103.081  106.099    10
```

File: R/data.table.R [hunk 3]
```diff
@@ -125,20 +125,17 @@ data.table <-function(..., keep.rownames=FALSE, check.names=FALSE, key=NULL, str
...
-        next
-      }
-      if (is.atomic(xi) || is.list(xi)) {
-        # TO DO: surely use set() here, or avoid the coercion
-        x[[i]] = rep(xi, length.out = nr)
-        next
-      }
-      stop("problem recycling column ",i,", try a simpler type")
-    # }
-    stop("argument ",i," (nrow ",nrows[i],") cannot be recycled without remainder to match longest nrow (",nr,")")
```
Comment (MichaelChirico) [inline R/data.table.R]: As near as I can tell this error is superseded by the above warning? Also eliminated the commented branch since it seems to be vestigial.

File: R/between.R [hunk 4]
```diff
@@ -22,18 +22,18 @@ inrange <- function(x,lower,upper,incbounds=TRUE) {
   subject = setDT(list(l=lower, u=upper))
   ops = if (incbounds) c(4L, 2L) else c(5L, 3L) # >=,<= and >,<
   verbose = getOption("datatable.verbose")
-  if (verbose) {last.started.at=proc.time()[3];cat("forderv(query) took ... ");flush.console()}
```
Comment (MichaelChirico) [inline R/between.R]: @mattdowle what about two functions, `timestart` and `timetaken`? `timestart` would accept a message as input, `cat` it, `flush.console()`s, and return the `proc.time()`; `timetaken` takes the result of `timestart`, `cat`s it, and `flush.console()`s, and returns nothing.

**Comment by mattdowle** [comment]
On the contrary, I'd say the `min` and `lq` figures confirm the 3%-4% improvement.  The other figures are affected too much by a varying number of `gc()`'s or effects like that. Given that the test suite has a large number of small arbitrary tests and we didn't think at all about an appropriate test, it's impressive any improvement shows up at all.  Take a particular test that hits those the changed lines, and repeat that one test many times and the improvement could be larger (but I'm not suggested doing that, no need). In R-devel the improvement may be greater by avoiding the small creation at all (which wasn't possible before as a type double).

**Comment by mattdowle** [comment]
Looks like it's a valid fail on test 1187.6 I see locally.  I'll leave to you.

File: R/data.table.R (hunk 3 above)
Comment (mattdowle) [reply R/data.table.R]: Agreed. +1

File: R/between.R (hunk 4 above)
Comment (mattdowle) [reply R/between.R]: Yes separate PR.  Sometimes there are several nested timing blocks in play.  Storing the `started.at` time in a variable and then passing that (or the appropriate one) to `timetaken` I think I like the balance of using a common function along with the flexibility of using it in different ways depending on the circumstances. i.e. current `timetaken()` approach, but actually using it here. IIUC.

**Comment by MichaelChirico** [comment]
Hmm, `test.data.table()` passes on my machine. All of the error logs truncate the error so I can't tell what the difference is here...

File: R/foverlaps.R [hunk 5]
```diff
@@ -128,24 +129,15 @@ foverlaps <- function(x, y, by.x = if (!is.null(key(x))) key(x) else key(y), by.
   }
   # nomatch has no effect here, just for passing arguments consistently to `bmerge`
   .Call(Clookup, uy, nrow(y), indices(uy, y, yintervals, nomatch=0L, roll=roll), maxgap, minoverlap, mult, type, verbose)
-  if (maxgap == 0L && minoverlap == 1L) {
```
Comment (MichaelChirico) [inline R/foverlaps.R]: @arunsrinivasan check out this diff. axed `iintervals` as it appears unused. And axed the extra argument checks to simplify so long as the other options are unimplemented

**Comment by MichaelChirico** [comment]
@mattdowle OK I think I've hammered away enough at the `codecov` diff miss for now. AFAICT the major remaining miss is from changes to `onAttach.R`, which I don't think (?) are testable.

File: R/foverlaps.R [hunk 6]
```diff
@@ -15,7 +15,7 @@ foverlaps <- function(x, y, by.x = if (!is.null(key(x))) key(x) else key(y), by.
   mult = match.arg(mult)
   if (type == "equal")
     stop("type = 'equal' is not implemented yet. But note that this is just the same as a normal data.table join y[x, ...], unless you are also interested in setting 'minoverlap / maxgap' arguments. But those arguments are not implemented yet as well.")
-  if (maxgap > 0L || minoverlap > 1L)
+  if (maxgap != 0L || minoverlap != 1L)
```
Comment (arunsrinivasan) [inline R/foverlaps.R]: maxgap < 0 and minoverlap < 1 both are checked in lines 6 and 8. What's the purpose of this change?

File: R/foverlaps.R [hunk 7]
```diff
@@ -104,10 +105,10 @@ foverlaps <- function(x, y, by.x = if (!is.null(key(x))) key(x) else key(y), by.
               end = yintervals[2L], any =,
               within =, equal = yintervals)
   call = construct(head(ynames, -2L), uycols, type)
-  if (verbose) {last.started.at=proc.time()[3];cat("unique() + setkey() operations done in ...");flush.console()}
+  if (verbose) {last.started.at=proc.time();cat("unique() + setkey() operations done in ...");flush.console()}
   uy = unique(y[, eval(call)])
-  setkey(uy)[, `:=`(lookup = list(list(integer(0))), type_lookup = list(list(integer(0))), count=0L, type_count=0L)]
-  if (verbose) {cat(round(proc.time()[3]-last.started.at,3),"secs\n");flush.console}
+  setkey(uy)[, `:=`(lookup = list(list(integer(0L))), type_lookup = list(list(integer(0L))), count=0L, type_count=0L)]
+  if (verbose) {cat(timetaken(last.started.at)); flush.console()}
```
Comment (arunsrinivasan) [inline R/foverlaps.R]: I'm a bit confused as to how these changes are related to this PR (1 -> 1L like enhancements). But I'm guessing you've discussed these changes with Matt.

File: R/foverlaps.R (hunk 5 above)
Comment (arunsrinivasan) [reply R/foverlaps.R]: @MichaelChirico good catch regarding `iintervals`. Seems like it wasn't used right from the first commit of `foverlaps`.

Comment (arunsrinivasan) [reply R/foverlaps.R]: I understand why you simplified code, but I left it there to serve as a placeholder when I'll come back to work on it... The first `if` statement would always be run currently. So that's the only cost here, which is fine.. Would it be possible to revert this part to how it was please?

The PR overall looks great! Apologies for not being able to write earlier.

File: R/foverlaps.R (hunk 7 above)
Comment (MichaelChirico) [reply R/foverlaps.R]: @arunsrinivasan the Codecov report was failing pretty miserably on the PR, since it seems to have uncovered a lot of un-tested lines (this one in particular would have come up from `3` becoming `3L`, for example

File: R/foverlaps.R (hunk 5 above)
Comment (MichaelChirico) [reply R/foverlaps.R]: @arunsrinivasan I figured that's why you left it like that... only changed the logic to make testing easier, Codecov was complaining since it's impossible to get to the other branches to test them (all hail the might Codecov 🌞  ⛪️ ). I figured it'd be easy enough to revert to the old code when needed (that's what git's for right?)

Comment (arunsrinivasan) [reply R/foverlaps.R]: Why not wrap those lines with `# nocov start` and `# nocov end` so as to skip coverage tests? See https://github.com/Rdatatable/data.table/blob/70db3e48738026388c6487e83ee902ad37256ca5/R/c.factor.R

Yes, commented out with a comment pointing to this PR would be okay too.

Comment (MichaelChirico) [reply R/foverlaps.R]: simply because i didn't know that was possible! I'll add that to `onAttach` and `onLoad` as well... thanks!

File: R/foverlaps.R [hunk 8]
```diff
@@ -130,26 +130,24 @@ foverlaps <- function(x, y, by.x = if (!is.null(key(x))) key(x) else key(y), by.
...
-  if (verbose) {cat(timetaken(last.started.at));flush.console}
-  olaps = .Call(Coverlaps, uy, xmatches, mult, type, nomatch, verbose)
+  if (maxgap == 0L && minoverlap == 1L) {
+    # iintervals = tail(names(x), 2L)    # iintervals not yet used so commented out for now
+    if (verbose) {last.started.at=proc.time();cat("binary search(es) done in ...");flush.console()}
+    xmatches = indices(uy, x, xintervals, nomatch=0L, roll=roll)
+    if (verbose) {cat(timetaken(last.started.at));flush.console()}
+    olaps = .Call(Coverlaps, uy, xmatches, mult, type, nomatch, verbose)
+  }
+  # nocov start
```
Comment (MichaelChirico) [inline R/foverlaps.R]: looks good

**Comment by MichaelChirico** [comment]
@mattdowle  thanks for the detailed review, i learned a lot about codecov here :)
